### PR TITLE
ROX-34929: Dedupe Scanner V4 Packages and Vulns

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Adds logic to dedupe packages and vulnerabilities in Scanner V4 results.

The deduping logic is disabled by default because it will be backported, to enable set `ROX_SCANNER_V4_DEDUPE: true`. 

### Deduping scenarios addressed

1. Duplicate RHCC packages (one per `kind`: `source`, `binary`, `ancestry`)
   - Only one is needed.
2. Duplicate vulns where only the advisory (RHSA) differs.
   - A single CVE may be fixed by multiple RHSA's, each RHSA would yield a 'duplicate' - now only the 'latest' is kept.
3. Different perspectives for the same vuln (ie: GHSA vs. GO, PYSEC vs. GHSA, etc.)
   - Logic added to pick the most complete / actionable entry

### Placement
Originally the logic was added to `mappers.go`, was later moved to `convert.go` as it was more readable and allowed for a 'rolling' approach when modifying the vuln details. In `mappers.go`, as vulns were combined new 'synthetic' vulns had to be created and added to the `vulnerabilities` map to not break other packages that may not reference the same set of duplicates (unnecessary complexity).

### Logic
Comparing by `severity` or `CVSS` was not enough, OSV/Red Hat have priority over NVD and in various scenarios the 'lower severity' needed to be top level.

To address known duplicate scenarios and keep data actionable, fields were grouped as 'fix' related and 'scoring' related (loose), the grouped fields are kept together such that the 'fix' info may come from duplicate A and the scoring from duplicate B.

| Scoring fields | Fix fields |
|---|---|
| cvss | advisory |
| cvssV2 | fixed by |
| cvssV3 | fix avail timestamp |
| scoreVersion | |
| link | |
| published\_on | |
| last\_modified | |
| severity | |
| summary | |

<details>
<summary><b>Deduping examples from real images</b></summary>

#### Scenario 1: Duplicate RHSAs — same CVE, different advisories

**Image:** `registry.redhat.io/compliance/openshift-compliance-openscap-rhel8@sha256:b64926c3cd92381b31f31f28ca01f05c52bc0fbbc581941ba6f0da5b3b142d36`
**Package:** `expat 2.5.0-3.el9_5.1` (rpm)

**Scoring:** equal — both have identical CVSS, severity, and summary.
**Fix:** B wins — newer advisory, newer fixed-by version, later fix timestamp.

| Field | Dupe A | Dupe B | Winner |
|---|---|---|---|
| CVE | CVE-2024-8176 | CVE-2024-8176 | — |
| Advisory | RHSA-2025:3531 | RHSA-2025:7444 | **B** (newer) |
| Fixed by | `0:2.5.0-3.el9_5.3` | `0:2.5.0-5.el9_6` | **B** (newer) |
| Fix timestamp | 2025-04-02 | 2025-05-13 | **B** (newer) |
| CVSS / severity | 7.5 / MODERATE | 7.5 / MODERATE | equal |

---

#### Scenario 2: Different perspectives — different CVSS scores

**Image:** `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:29e3563d5fc11a54ffbf7b28694f1cc3975a72e24035fca4afa44d7937e932c7`
**Package:** `github.com/docker/docker v27.5.1+incompatible` (Go binary)

**Scoring:** A wins — has 2 CVSS metric sources vs 1.
**Fix:** equal — neither has an advisory or fix version.

| Field | Dupe A | Dupe B | Winner |
|---|---|---|---|
| CVE | CVE-2026-33997 | CVE-2026-33997 | — |
| CVSS | 6.8 | 8.1 | **A** (more sources) |
| CVSS vector | `AV:N/AC:H/...` | `AV:N/AC:L/...` | **A** |
| Severity | MODERATE | IMPORTANT | **A** |
| Link | osv.dev | nvd.nist.gov | **A** |
| Published | 2026-03-27 | 2026-04-02 | **A** (earlier) |
| CVSS metrics | 2 sources | 1 source | **A** |

---

#### Scenario 3: Different perspectives — one has fix version

**Same image as Scenario 2**
**Package:** `github.com/docker/docker v27.5.1+incompatible` (Go binary)

**Scoring:** A wins — higher CVSS (8.8 vs 7.8), 2 metric sources vs 1.
**Fix:** A wins — has fix version `29.3.1`; B has none.

| Field | Dupe A | Dupe B | Winner |
|---|---|---|---|
| CVE | CVE-2026-34040 | CVE-2026-34040 | — |
| CVSS | 8.8 | 7.8 | **A** |
| Fixed by | `29.3.1` | empty | **A** |
| Severity | IMPORTANT | IMPORTANT | equal |
| Link | osv.dev | nvd.nist.gov | **A** |
| CVSS metrics | 2 sources | 1 source | **A** |

---

#### Scenario 4: Different perspectives — one has no scoring at all

**Same image as Scenario 2**
**Package:** `github.com/go-jose/go-jose/v3 v3.0.4` (Go binary)

**Scoring:** A wins — has CVSS, severity, and metric source; B has none.
**Fix:** equal — both have `3.0.5` as the fix version.

| Field | Dupe A | Dupe B | Winner |
|---|---|---|---|
| CVE | CVE-2026-34986 | CVE-2026-34986 | — |
| CVSS | 7.5 | empty | **A** |
| CVSS v3 | full vector | empty | **A** |
| Severity | IMPORTANT | UNKNOWN | **A** |
| Fixed by | `3.0.5` | `3.0.5` | equal |
| CVSS metrics | 1 source | empty | **A** |

---

#### Scenario 5: Different perspectives — different CVSS severity

**Same image as Scenario 2**
**Package:** `setuptools 53.0.0` (Python)

**Scoring:** A wins — higher CVSS (7.5 vs 5.9), 2 metric sources vs 1.
**Fix:** equal — both have `65.5.1` as the fix version.

| Field | Dupe A | Dupe B | Winner |
|---|---|---|---|
| CVE | CVE-2022-40897 | CVE-2022-40897 | — |
| CVSS | 7.5 | 5.9 | **A** |
| CVSS vector | `AV:N/AC:L/...` | `AV:N/AC:H/...` | **A** |
| Severity | IMPORTANT | MODERATE | **A** |
| Link | osv.dev | nvd.nist.gov | **A** |
| Fixed by | `65.5.1` | `65.5.1` | equal |
| CVSS metrics | 2 sources | 1 source | **A** |

</details>

## User-facing documentation

- [x] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [x] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [x] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [x] modified existing tests

### How I validated my change

unit tests + CI + manually

Using a custom dupe detector and two different infra clusters, one with the fix the other without, analyzed each active image for duplicates.

The cluster without the fix:

```
Summary
  Images scanned:          96
  Total components:        26758
  Total CVE entries:       7656
  Duplicate CVEs:          436
  Identical components:    370

  If duplicates removed (estimate):
    CVEs:        7656 -> 7220  (5.7% reduction)
    Components:  26758 -> 26388  (1.4% reduction)
```

```
Summary
  Images scanned:          96
  Total components:        26387
  Total CVE entries:       7220
  Duplicate CVEs:          0
  Identical components:    0
```

### CODE REVIEW

**Comment by dcaravel** [comment]
@coderabbitai review

**Review by vikin91 — COMMENTED** [review]
Despite not being in the Scanner-core, I gave it a selective look. My human eyes didn't catch much, but I threw it at AI and there are two things that you may want to check:

1. Question: 

does this layer intentionally rely on `PackageVulnerabilities` already being advisory-deduped upstream when `ROX_SCANNER_V4_RED_HAT_CVES=false`?

I noticed `vulnerabilities()` now merges by `ccVuln.GetName()`. In isolation that would collapse two entries like:
```
- RHSA-2024:10775 -> ...CVE-2024-24789...
- RHSA-2024:10775 -> ...CVE-2024-24790...
```
into one result.

I think `toProtoV4PackageVulnerabilitiesMap()` / `dedupeAdvisories()` currently makes that unreachable, so this may be fine. Just wanted to confirm that this function is deliberately depending on that upstream invariant.


2. Potential issue:

I think `mergeFixFields()` may regress remediation data in a real way, not just reshuffle metadata. Could we add a regression test for the case where a newer advisory has no `FixedInVersion`, but an older duplicate does?
Something like:

```go
t.Run("newer advisory does not erase existing fix version", func(t *testing.T) {
	vulnMap := map[string]*v4.VulnerabilityReport_Vulnerability{
		"a": {
			Id: "a", Name: "CVE-2024-1",
			Advisory:       &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:100"},
			FixedInVersion: "1.0.0",
		},
		"b": {
			Id: "b", Name: "CVE-2024-1",
			Advisory: &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:200"},
		},
	}

	got := vulnerabilities(vulnMap, []string{"a", "b"}, "", "")
	require.Len(t, got, 1)

	// Prefer the newer advisory metadata, but do not lose known fixability.
	assert.Equal(t, "RHSA-2024:200", got[0].GetAdvisory().GetName())
	assert.Equal(t, "1.0.0", got[0].GetFixedBy())
})
```
This would help answer whether losing fixability here is really intended.

**Comment by dcaravel** [comment]
> 1. Question:
> 
> does this layer intentionally rely on `PackageVulnerabilities` already being advisory-deduped upstream when `ROX_SCANNER_V4_RED_HAT_CVES=false`?
> 
> I noticed `vulnerabilities()` now merges by `ccVuln.GetName()`. In isolation that would collapse two entries like:
> 
> ```
> - RHSA-2024:10775 -> ...CVE-2024-24789...
> - RHSA-2024:10775 -> ...CVE-2024-24790...
> ```
> 
> into one result.
> 
> I think `toProtoV4PackageVulnerabilitiesMap()` / `dedupeAdvisories()` currently makes that unreachable, so this may be fine. Just wanted to confirm that this function is deliberately depending on that upstream invariant.

That's correct, by the time the the information gets to Central there should only be a single RHSA, and those RHSAs do not map to a single CVE, they generally represent many CVEs - the behavior in dedupeAdvisories addresses that.

Regardless - ROX_SCANNER_V4_RED_HAT_CVES was to be removed when the dedicated advisory field was added (ROX-26672) but never was. I wouldn't put too much focus on that path, but either way it should behave OK.



> 2. Potential issue:
> 
> I think `mergeFixFields()` may regress remediation data in a real way, not just reshuffle metadata. Could we add a regression test for the case where a newer advisory has no `FixedInVersion`, but an older duplicate does? Something like:
> 
> ```go
> t.Run("newer advisory does not erase existing fix version", func(t *testing.T) {
> 	vulnMap := map[string]*v4.VulnerabilityReport_Vulnerability{
> 		"a": {
> 			Id: "a", Name: "CVE-2024-1",
> 			Advisory:       &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:100"},
> 			FixedInVersion: "1.0.0",
> 		},
> 		"b": {
> 			Id: "b", Name: "CVE-2024-1",
> 			Advisory: &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:200"},
> 		},
> 	}
> 
> 	got := vulnerabilities(vulnMap, []string{"a", "b"}, "", "")
> 	require.Len(t, got, 1)
> 
> 	// Prefer the newer advisory metadata, but do not lose known fixability.
> 	assert.Equal(t, "RHSA-2024:200", got[0].GetAdvisory().GetName())
> 	assert.Equal(t, "1.0.0", got[0].GetFixedBy())
> })
> ```
> 
> This would help answer whether losing fixability here is really intended.

That is the opposite of the existing test case:

```go
t.Run("advisory wins with no fix version", func(t *testing.T) {
		vulnMap := map[string]*v4.VulnerabilityReport_Vulnerability{
			"a": {
				Id: "a", Name: "CVE-2024-1",
				Advisory:       &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:100"},
				FixedInVersion: "1.0.0",
			},
			"b": {
				Id: "b", Name: "CVE-2024-1",
				Advisory: &v4.VulnerabilityReport_Advisory{Name: "RHSA-2024:200"},
			},
		}
		got := vulnerabilities(vulnMap, []string{"a", "b"}, "", "")
		require.Len(t, got, 1)
		// "b" wins on advisory — fix version is empty from "b".
		assert.Equal(t, "RHSA-2024:200", got[0].GetAdvisory().GetName())
		assert.Equal(t, "", got[0].GetFixedBy())
	})
```
This scenario shouldn't occur - an RHSA describes a 'fix', any versions prior to the version mentioned are considered vulnerable, an RHSA wouldn't be published for a package it doesn't fix.

But lets say that can happen - if we surface to the user that this vulnerability is fixed in 1.0.0, but a later RHSA says the package is no longer fixed (not possible), that user will still be affected by the vuln in that case.  I could argue that 'no fix' is the more accurate conclusion (which is why I created that test / expected outcome)

If we feel that we should carry the fix version forward in this case, I'm OK changing it - either way should not occur in practice.

**Comment by vikin91** [comment]
Thank you for answering! The answers make sense! 
I can mark the PR green, but because I am not 100% in the topic I would suggest that maybe someone else should give this a look too.

**Review by vikin91 — APPROVED** [review]
Approving as in my comment (suggesting that second pair of eyes give that a look, as I may lack some assumptions).

File: pkg/scanners/scannerv4/convert.go [hunk 1]
```diff
@@ -48,8 +55,37 @@ func componentsWithLayerMap(metadata *storage.ImageMetadata, report *v4.Vulnerab
 			pkgs[pkg.GetId()] = pkg
 		}
 	}
+	// Filter out non-binary packages that should not become user-facing components:
```
Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go]: - I'm not sure "ancestry carries VEX suppression metadata only" is accurate. Aren't they providing layer-based component specification? They tell us "these layers are 'product/component "foobar"'. Yes, we are using them to suppress "known not affected", but that's not what they semantically mean.
- I'm OK keeping "source" for now, but, referencing them in results in the case we flag vulnerabilities for them will likely be irrelevant since "source" packages are never installed and my understanding is that they provide "indirect" vulnerability matching mechanism in the security data.

Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go]: Pedantic nit: Remove ancestry packages don't seem related to "deduping" in general.

File: pkg/scanners/scannerv4/convert.go [hunk 2]
```diff
@@ -587,3 +648,142 @@ func notes(report *v4.VulnerabilityReport) []storage.ImageScan_Note {
...
+		}
+		return -1
+	}
+	// Reaching here means both have a fix, neither matches pkgFixedBy (or it
+	// is empty), and the versions disagree. This is rare — it requires two
+	// sources to report different fix versions for the same CVE. Use a
+	// deterministic numeric comparison so the result is stable across runs.
+	if a != b {
+		return compareNumericSegments(a, b)
+	}
```
Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go:696]: I want to ack this is "best-effort" and we might not get the actual latest fixed by, because we might not know how to compare the versions schema.

Can we log a warning when we get to `a != b`, for visibility? Maybe debug, if it's common (but your comment says "rare").

File: pkg/scanners/scannerv4/convert.go [hunk 3]
```diff
@@ -587,3 +648,142 @@ func notes(report *v4.VulnerabilityReport) []storage.ImageScan_Note {
...
+	if a == nil && b == nil {
+		return 0
+	}
+	if a == nil {
+		return -1
+	}
+	if b == nil {
+		return 1
+	}
+	return compareNumericSegments(a.GetName(), b.GetName())
```
Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go]: A likely rare situation, but we don't care if the advisory is RHSA, RHBA, RHEA, etc, right?

File: pkg/scanners/scannerv4/convert.go [hunk 4]
```diff
@@ -587,3 +648,142 @@ func notes(report *v4.VulnerabilityReport) []storage.ImageScan_Note {
...
 }
+
+// mergeFixFields overwrites fix-related fields on dst when src has more
+// recent or more complete fix data. Priority: later advisory, has fix over
+// doesn't, matches package-level fix version, higher version by numeric
+// comparison.
+func mergeFixFields(dst *storage.EmbeddedVulnerability, src *v4.VulnerabilityReport_Vulnerability, envOS, pkgFixedByVersion string) {
+	srcAdv := advisory(src.GetAdvisory())
+	c := cmp.Or(
+		compareAdvisories(srcAdv, dst.GetAdvisory()),
```
Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go]: Advisories win, but I want to ask the impact of a newer advisory not having a fixed version, and we overwrite results that previously said "you can fix this by upgrading to X.Y.Z".

**Review by jvdm — COMMENTED** [review]
Would it be simpler to `buildEmbeddedVulnerability` for the dedup candidate first, then have the merge functions compare two `*storage.EmbeddedVulnerability` values? The merge functions re-implement parts of `buildEmbeddedVulnerability()` (`advisory()`, `vulnDataSource()`, `link()`...) and nothing guarantees these two paths stay in sync. There might be bugs there.

File: pkg/scanners/scannerv4/convert.go (hunk 4 above)
Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go]: By the way, I saw an earlier answer in the PR:

> But lets say that can happen - if we surface to the user that this vulnerability is fixed in 1.0.0, but a later RHSA says the package is no longer fixed (not possible), that user will still be affected by the vuln in that case. I could argue that 'no fix' is the more accurate conclusion (which is why I created that test / expected outcome)

For RHSA it's fine, but for different authorities talking about the same package (eg. GO and GHSA), this could surface, right?

Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go]: So, my question is not on correctness, but rather, understanding how users will experience the same CVE having a fixed version and not having one anymore (let's say GO is more "accurate" than GHSA), and if the impact justify us making an exception for that scenario or not.

Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go]: OK, I think I managed to answer my own question: The advisory function in mappers.go explicitly ignores anything that isn't a Red Hat updater (vuln.Updater != RedHatUpdaterName). Only RHSA/RHBA/RHEA patterns are matched. So, the scenario is more of a corner case right now since RHSA/RHBA/RHEA are supposed to have a fix. You can resolve this if my own thought process looks sane to you.

File: pkg/scanners/scannerv4/convert.go (hunk 1 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go]: > I'm not sure "ancestry carries VEX suppression metadata only" is accurate. Aren't they providing layer-based component specification? They tell us "these layers are 'product/component "foobar"'. Yes, we are using them to suppress "known not affected", but that's not what they semantically mean.

Removed the details from the comment to make it more general / accurate.

> I'm OK keeping "source" for now, but, referencing them in results in the case we flag vulnerabilities for them will likely be irrelevant since "source" packages are never installed and my understanding is that they provide "indirect" vulnerability matching mechanism in the security data.

👍 

Was [this logic](https://github.com/quay/claircore/blob/main/datastore/postgres/querybuilder.go#L28) in claircore that made me believe a top level `source` package is possible.

File: pkg/scanners/scannerv4/convert.go (hunk 3 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go]: Thought about this more - changed this to compare 'FixedDate' instead which is a stronger signal. Every advisory has a fixed date, and we only populate fixed date for RH advisories. This way we grab the latest regardless of type, which was the intent.

File: pkg/scanners/scannerv4/convert.go (hunk 1 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go]: Hope I understood this correctly - I made the comment more general. Let me know otherwise.

File: pkg/scanners/scannerv4/convert.go (hunk 3 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go]: Had the robots run some numbers on the Scanner V4 data

> 79 out of 474,090 distinct rhel-vex CVE+package combinations (0.02%) have mixed advisory types

**Review by jvdm — APPROVED** [review]
